### ISSUE TITLE
Doorbell event support for EZVIZ DB-1|Hikvision DS-HD1 & other models

### ISSUE THREAD

**Issue by bgilmer77** [Skinah/IpCamera#56]
This is a feature request to add event support for the EZVIZ DB-1 video doorbell running Hikvision 200321 Firmware.  A *lot* of information on this doorbell may be found [here](https://ipcamtalk.com/threads/new-rca-hsdb2a-3mp-doorbell-ip-camera.31601/page-101).

Here are a few user stories

- As a user who is familiar with OpenHAB2 home automation, I want to be able to capture motion events from the PIR sensor on my doorbell so that I can detect motion in front of this camera without having to increase the load on my OH2 server by doing ffmpeg motion analysis on the OH2 server.
- As a user who knows how to set up motion detection such as line crossing using HikVision software, I want to be able to receive motion event notification in OH2 from the HikVision software so that I can get the full benefits of HikVision motion analysis in my OH2 implementation.
-  As a user who knows how to configure rules in OH2, I want to be able to create rules based on both PIR and camera-software reported events so that I can fine-tune motion detection and reduce false alarms.

I will use this issue to upload information which I hope will be helpful in identifying current issues with the ipcamera binding, and aid in being able to access events from this doorbell camera.

**Comment by bgilmer77** [comment]
**System description**
Here is a description of the hardware and software associated with this issue
_OpenHAB system information_
OH version - 2.5.5 Release Build
OS version - Ubuntu 18.04.4 LTS Desktop (bionic)
Hardware platform - Intel NUC NUC7i5BN, dual core, 16 GB memory, 300GB flash drive

_Doorbell information_
EZVIZ Model CD- DB1 (A0-1B3WPFR)
Firmware - varies and will be noted in README uploaded with any associated files.

**Comment by bgilmer77** [comment]
There is a utility found at [https://github.com/dmitrif/onvif2mqtt](https://github.com/dmitrif/onvif2mqtt) which successfully connects to this doorbell.  The utility is able to create an ONVIF pullpoint subscription, and then get notifications of PIR events using ONVIF.  This utility works with my doorbell, but ONLY running LaView firmware.  It DOES NOT work with HikVision firmware.  The following is from the Hardware Compatibility section of Dimitri's README.

- EZViz DB1 Doorbell (flashed with LaView firmware) - **TESTED**
- Nelly's Security Doorbell (NSC-DB2)
- Laview Halo One Doorbell - **TESTED**
- RCA HSDB2A Doorbell (flashed with LaView firmware)
- Hikvision DB2 - **Does not work, TODO: implement push point subscriptions.**
- Any other ONVIF compliant IP Camera - if it works for you please let me know so that this list can be updated.

I have used wireshark to capture network traffic as this utility creates a pullpoint event subscription on the camera.  **The camera is running LaView FW**.  The file name of the capture is `onvif2mqtt_connecting_create_pullpint_event_subscription_laveiwFW.pcapng`

Here is the output of the utility as the subscription was created.
```
root@OH-Test /opt/onvif2mqtt # npm run dev

> onvif-cam@1.0.4 dev /opt/onvif2mqtt
> cross-env CONFIG_PATH=./config.dev.yml nodemon -e js,yml --exec babel-node src/index.js | pino-pretty

[nodemon] 2.0.2
[nodemon] to restart at any time, enter `rs`
[nodemon] watching dir(s): *.*
[nodemon] watching extensions: js,yml
[nodemon] starting `babel-node src/index.js`
[1590358701864] INFO : Loading configuration. {"configPath":"./config.dev.yml"}
[1590358701869] INFO : Validating configuration file.
[1590358701953] INFO  (Manager): Beginning initialization...
[1590358701953] INFO  (MQTT on 127.0.0.1:1883): Connecting.
[1590358701983] INFO  (MQTT on 127.0.0.1:1883): Successfully connected.
[1590358701984] INFO  (ONVIF/doorbell on 192.168.18.30): Attempting connection.
[1590358701985] DEBUG (MQTT on 127.0.0.1:1883): Publishing. {"topic":"onvif2mqtt/doorbell/motion","value":"OFF","retain":true}
[1590358702149] INFO  (ONVIF/doorbell on 192.168.18.30): Successfully connected.
```

I have another capture during two events, a PIR ON and a PIR off.  Here is the output of the utility as the PIR events were captured.
```
[1590358886646] DEBUG (MQTT on 127.0.0.1:1883): Publishing. {"topic":"onvif2mqtt/doorbell/motion","value":"ON","retain":true}
[1590358892624] DEBUG (MQTT on 127.0.0.1:1883): Publishing. {"topic":"onvif2mqtt/doorbell/motion","value":"OFF","retain":true}
```

To sum up, this is an example of a working connection, the creation of a pull point subscription, and the subsequent transmission of two events using that subscription.

However, this is NOT the firmware I would like to use with this doorbell.  The LaView firmware is very simple and does not contain any of the software-based event notification contained in the HikVision firmware.  Ad Dimitri mentions in his GH README, he has a TODO to implement push point subscriptions.  I did a little digging and did not find such a thing as push point subscriptions in the ONVIF documentation.  Perhaps this is unique to HikVision?  Perhaps Dimitri is wrong?  Don't know.
[onvif2mqtt_connecting_create_pullpoint_event_subscription_laviewFW.zip](https://github.com/Skinah/IpCamera/files/4678388/onvif2mqtt_connecting_create_pullpoint_event_subscription_laviewFW.zip)

[onvif2mqtt_PIR_event_ON_OFF_laviewFW.zip](https://github.com/Skinah/IpCamera/files/4678389/onvif2mqtt_PIR_event_ON_OFF_laviewFW.zip)



- EZViz DB1 Doorbell (flashed with LaView firmware) - **TESTED**
- Nelly's Security Doorbell (NSC-DB2)
- Laview Halo One Doorbell - **TESTED**
- RCA HSDB2A Doorbell (flashed with LaView firmware)
- Hikvision DB2 - **Does not work, TODO: implement push point subscriptions.**
- Any other ONVIF compliant IP Camera - if it works for you please let me know so that this list can be updated.

I have used wireshark to capture this utility creating a pullpoint event subscription, and I have captured this same utility as two events, a PIR ON and a PIR off, were

**Comment by bgilmer77** [comment]
**EZVIZ DB-1 with EZVIZ FW (191211)**
The EZVIZ FW does not support ONVIF.  ODM will not connect to the doorbell even though it is alive on the network and can be accessed using the EZVIZ app.

onvif2mqtt utility produces the following dialog when attempting to connect.
```
root@OH-Test /opt/onvif2mqtt # npm run dev

> onvif-cam@1.0.4 dev /opt/onvif2mqtt
> cross-env CONFIG_PATH=./config.dev.yml nodemon -e js,yml --exec babel-node src/index.js | pino-pretty

[nodemon] 2.0.2
[nodemon] to restart at any time, enter `rs`
[nodemon] watching dir(s): *.*
[nodemon] watching extensions: js,yml
[nodemon] starting `babel-node src/index.js`
[1590432254810] INFO : Loading configuration. {"configPath":"./config.dev.yml"}
[1590432254816] INFO : Validating configuration file.
[1590432254896] INFO  (Manager): Beginning initialization...
[1590432254897] INFO  (MQTT on 127.0.0.1:1883): Connecting.
[1590432254925] INFO  (MQTT on 127.0.0.1:1883): Successfully connected.
[1590432254925] INFO  (ONVIF/doorbell on 192.168.18.30): Attempting connection.
[1590432254926] DEBUG (MQTT on 127.0.0.1:1883): Publishing. {"topic":"onvif2mqtt/doorbell/motion","value":"OFF","retain":true}
[1590432254932] ERROR (ONVIF/doorbell on 192.168.18.30): Failed to connect to doorbell {"errno":-111,"code":"ECONNREFUSED","syscall":"connect","address":"192.168.18.30","port":80}
[1590432254932] INFO  (ONVIF/doorbell on 192.168.18.30): Successfully connected.
/opt/onvif2mqtt/node_modules/onvif/lib/cam.js:185
			? (this.uri[options.service] ? this.uri[options.service].path : options.service)
			           ^

TypeError: Cannot read property 'events' of undefined
    at Cam._request (/opt/onvif2mqtt/node_modules/onvif/lib/cam.js:185:15)
    at Cam.createPullPointSubscription (/opt/onvif2mqtt/node_modules/onvif/lib/events.js:86:8)
    at Cam._eventRequest (/opt/onvif2mqtt/node_modules/onvif/lib/events.js:250:9)
    at Cam.<anonymous> (/opt/onvif2mqtt/node_modules/onvif/lib/events.js:237:9)
    at Cam.emit (events.js:315:20)
    at Cam.EventEmitter.emit (domain.js:485:12)
    at _addListener (events.js:358:14)
    at Cam.on [as addListener] (events.js:406:10)
    at Cam.<anonymous> (/opt/onvif2mqtt/src/onvif/Subscriber.js:38:14)
    at Cam.<anonymous> (/opt/onvif2mqtt/node_modules/onvif/lib/cam.js:106:20)
[nodemon] app crashed - waiting for file changes before starting...
```

**Comment by bgilmer77** [comment]
**EZVIZ DB-1 with HikVision FW (200321)**

Summary:
-  Both ODM and onvif2mqtt successfully connect and report PIR events with this configuration
-  ODM connects to the device
-  ODM shows PIR events
-  onvif2mqtt successfully connects, creates pull point subscription, and reports PIR events
-  Wireshark capture of onvif2mqtt connection session shows in line 35 that both Subscription Policy Support is True, and that Pull Point Support is True (response in 'event' section to capability query)
-  Same wireshark capture shows in line 66 a response to  GetProfiles quiry with video and audio format information
-  Same wireshark capture shows in line 82, a request to CreatePullPointSubscription.  Line 89 returns OK.  See https://www.onvif.org/onvif/ver10/events/wsdl/event.wsdl
-  Same wireshark capture shows in line 96, authenticated login and interaction with PullSubManager setting some parameters.

[onvif2mqtt_connecting_create_pullpoint_event_subscription_hikvisionFW.zip](https://github.com/Skinah/IpCamera/files/4678871/onvif2mqtt_connecting_create_pullpoint_event_subscription_hikvisionFW.zip)

A wireshark capture of event notifications shows the following.
- Line 14, camera issues PullMessageResponse (fwiw, I could find nothing about this in the ONVIF documentation)
-  Line 20, onvif2mqtt issues 'PullMessages' with some parameters set.
-  Line 32, camera responds with information about a motion event

[onvif2mqtt_PIR_event_ON_OFF_hikvisionFW.zip](https://github.com/Skinah/IpCamera/files/4678872/onvif2mqtt_PIR_event_ON_OFF_hikvisionFW.zip)

ODM finds the doorbell.  It also shows PIR events, as shown in this screen capture.
![ODM_PIR_event_hikvisionFW](https://user-images.githubusercontent.com/7648629/82843847-7f911700-9eac-11ea-8094-0da047682365.png)

**Comment by bgilmer77** [comment]
** OH2 using EZVIZ DB-1 with HikVision Firmware**
In this sequence, I had a .things file ready to go, but with the wrong name (ipcamera.things~).  I started OH2 and created the following capture with wireshark.
[oh2_startup_hikvisionFW.zip](https://github.com/Skinah/IpCamera/files/4679239/oh2_startup_hikvisionFW.zip)

Once OH2 was up and running, I renamed the .things file so that it was picked up by OH2.  Here is the .things file.
```
Thing ipcamera:ONVIF:doorbell
[
	IPADDRESS="192.168.18.30",
	USERNAME="admin",
	PASSWORD="JEGRAK",
	POLL_CAMERA_MS=2000000,
	SNAPSHOT_URL_OVERRIDE="http://192.168.18.30/onvif/snapshot.jpg",
	SERVER_PORT=54030,
	FFMPEG_INPUT="rtsp://192.168.18.30:554/Streaming/Channels/101",
	FFMPEG_OUTPUT="/tmpfs/doorbell/"
]
```
Here is a wireshark capture from when I renamed the .things file.
[oh2_add_ipcamera.things_hikvisionFW.zip](https://github.com/Skinah/IpCamera/files/4679247/oh2_add_ipcamera.things_hikvisionFW.zip)

Once the doorbell camera was defined as a thing, I generated a PIR motion event.  This is what I captured on wireshark.
[oh2_pir_event_hikvisionFW.zip](https://github.com/Skinah/IpCamera/files/4679251/oh2_pir_event_hikvisionFW.zip)
To me, it looks like everything worked.  The doorbell called up OH and reported a motion alarm.

Here is the .items file.
```
String doorbell_MJPEGStreamURL "MJPEG URL" { channel="ipcamera:ONVIF:doorbell:streamUrl" }
Switch doorbellMotionAlarm "Motion Detected" { channel="ipcamera:ONVIF:doorbell:motionAlarm" }
Dimmer doorbellMotionThreshold "Motion Threshold" { channel="ipcamera:ONVIF:doorbell:ffmpegMotionControl" }
```
Here is the .sitemaps file
```
sitemap ipcamera label="IP Cameras" {

	Frame label="Doorbell Camera" {
		Text item= doorbell_MJPEGStreamURL
		Text label="Snapshot" icon="camera"{Image url="http://192.168.18.5:54030/ipcamera.jpg"}
		Text label="MJPEG Stream" icon="camera"{Video url="http://192.168.18.5:54030/ipcamera.mjpeg" encoding="mjpeg"}
		Switch item= doorbellMotionAlarm
		Slider item= doorbellMotionThreshold
	}
}
```
I watched the "Motion Alarm" switch item as I tripped the PIR, but the switch never changed state.

Here is the openhab.log file.
[openhab.zip](https://github.com/Skinah/IpCamera/files/4679258/openhab.zip)

Here is the events.log file.
[events.zip](https://github.com/Skinah/IpCamera/files/4679259/events.zip)

I think that is it for now.  Please let me know if you need anything else.

**Comment by Skinah** [comment]
events/log is useless as it only shows if the binding was working :) Dont bother sending that log.

openhab.log showed it did not pass authentication so it was useless, possibly a timezone / time /date did not match...

oh2_pir_event_hikvisionFW.zip
That file was very useful and showed I was very close to having it work, this probably gave me what I needed.


I think your in luck.

GetServicesResponse
returns this
<tds:Service><tds:Namespace>http://www.onvif.org/ver10/events/wsdl</tds:Namespace>
<tds:XAddr>http://192.168.18.30/onvif/Events</tds:XAddr>

Which is what the Onvif library is using, my Hikvision sends it to a totally different namespace/address so this means I will need to reply on you for testing as my camera wont work.

**Comment by bgilmer77** [comment]
>openhab.log showed it did not pass authentication so it was useless, possibly a timezone / time /date did not match...

FWIW, I had seen that having time not match is an issue with authentication.  I have an NTP server running on the OH2 box.  The camera sets its time from the OH2 NTP server, so they both should have the same time.

Getting ready to try out your latest Alpha.

**Comment by bgilmer77** [comment]
- Loaded 05-26 alpha
- revised .items and .sitemap to include new PIR motion switch
- restarted OH2 three times
- Generated PIR event while running wireshark
- No indication of change of state of doorbellPIRAlarm in events.log or by change in state of switch item in sitemap

Wireshark capture attached.
[2020_05_26_oh2_pir_event_hikvisionFW.zip](https://github.com/Skinah/IpCamera/files/4683081/2020_05_26_oh2_pir_event_hikvisionFW.zip)

**Comment by Skinah** [comment]
I really need the debug level logs from openhab.log it gives me info that wireshark can not see. It is also possible this will stop working 1 minute after the binding starts up as I have not implemented the renew process to keep new messages coming in so very important to be a working startup of the binding that is shown in the log file.

**Comment by bgilmer77** [comment]
Standby.  I will collect and post the OH debug log here in a few minutes.

**Comment by bgilmer77** [comment]
Started OH2.  Waited about 45 seconds.  Generated PIR event.  Stopped OH2.
[openhab.log](https://github.com/Skinah/IpCamera/files/4684856/openhab.log)

**Comment by Skinah** [comment]
Did you turn everything to debug? Please leave all on INFO and only set the ipcamera binding to DEBUG. The readme has how to do this if you need the console command.

**Comment by bgilmer77** [comment]
I do not know why that log contained so much spam.  I am familiar with the log console command.  Anyway, here is a new log file.  I found that ROOT was spamming the log, so I did a log:set ERROR ROOT.

Unfortunately, with the binding set to DEBUG or TRACE, there were no new log entries during a PIR event.

I did two tests.  Test #1 was with the log level for the binding set to DEBUG.  Here is the file.
[openhab.DEBUG.log](https://github.com/Skinah/IpCamera/files/4685463/openhab.DEBUG.log)

Then I set the binding to TRACE.  Here is the log.
[openhab.TRACE.log](https://github.com/Skinah/IpCamera/files/4685466/openhab.TRACE.log)

Here is what log:list showed after the first test.
```
openhab> log:list
Logger                                             │ Level
───────────────────────────────────────────────────┼──────
ROOT                                               │ ERROR
javax.jmdns                                        │ ERROR
javax.mail                                         │ ERROR
org.apache.karaf.jaas.modules.audit                │ INFO
org.apache.karaf.kar.internal.KarServiceImpl       │ ERROR
org.apache.karaf.shell.ssh.SshUtils                │ ERROR
org.apache.karaf.shell.support                     │ OFF
org.eclipse.lsp4j                                  │ OFF
org.eclipse.smarthome                              │ INFO
org.jupnp                                          │ ERROR
org.openhab                                        │ INFO
org.openhab.binding.ipcamera                       │ DEBUG
org.openhab.ui.paper                               │ WARN
org.openhab.ui.paper.internal                      │ INFO
org.ops4j.pax.url.mvn.internal.AetherBasedResolver │ ERROR
org.ops4j.pax.web.pax-web-runtime                  │ OFF
smarthome.event                                    │ INFO
smarthome.event.InboxUpdatedEvent                  │ ERROR
smarthome.event.ItemAddedEvent                     │ ERROR
smarthome.event.ItemRemovedEvent                   │ ERROR
smarthome.event.ItemStateEvent                     │ ERROR
smarthome.event.ThingAddedEvent                    │ ERROR
smarthome.event.ThingRemovedEvent                  │ ERROR
smarthome.event.ThingStatusInfoEvent               │ ERROR
```
Procedure was to delete openhab.log, create a new openhab.log.  Start OH.  Wait 30 seconds.  Generate PIR event.  Stop OH.

**Comment by Skinah** [comment]
Thank you for the cleaner log and for doing the captures. Wont get time to look at them in detail for a day or two.

Your openhab Time zone set to 'Canada/Eastern'.
Can you login with Onvif device manager and check the camera is set to the same timezone ?

The below is going to block the binding until it is resolved hence why this is a warning level. None of the new code is running so I have no idea where your wireshark traces are getting the traffic from as the binding is not the trigger for them...


2020-05-26 19:16:50.419 [WARN ] [ing.ipcamera.handler.IpCameraHandler] - ONVIF error 400:<?xml version="1.0" encoding="UTF-8"?>
<env:Envelope xmlns:env="http://www.w3.org/2003/05/soap-envelope" xmlns:soapenc="http://www.w3.org/2003/05/soap-encoding" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xmlns:xs="http://www.w3.org/2001/XMLSchema" xmlns:tt="http://www.onvif.org/ver10/schema" xmlns:tds="http://www.onvif.org/ver10/device/wsdl" xmlns:trt="http://www.onvif.org/ver10/media/wsdl" xmlns:timg="http://www.onvif.org/ver20/imaging/wsdl" xmlns:tev="http://www.onvif.org/ver10/events/wsdl" xmlns:tptz="http://www.onvif.org/ver20/ptz/wsdl" xmlns:tan="http://www.onvif.org/ver20/analytics/wsdl" xmlns:tst="http://www.onvif.org/ver10/storage/wsdl" xmlns:ter="http://www.onvif.org/ver10/error" xmlns:dn="http://www.onvif.org/ver10/network/wsdl" xmlns:tns1="http://www.onvif.org/ver10/topics" xmlns:tmd="http://www.onvif.org/ver10/deviceIO/wsdl" xmlns:wsdl="http://schemas.xmlsoap.org/wsdl" xmlns:wsoap12="http://schemas.xmlsoap.org/wsdl/soap12" xmlns:http="http://schemas.xmlsoap.org/wsdl/http" xmlns:d="http://schemas.xmlsoap.org/ws/2005/04/discovery" xmlns:wsadis="http://schemas.xmlsoap.org/ws/2004/08/addressing" xmlns:xop="http://www.w3.org/2004/08/xop/include" xmlns:wsnt="http://docs.oasis-open.org/wsn/b-2" xmlns:wsa="http://www.w3.org/2005/08/addressing" xmlns:wstop="http://docs.oasis-open.org/wsn/t-1" xmlns:wsrf-bf="http://docs.oasis-open.org/wsrf/bf-2" xmlns:wsntw="http://docs.oasis-open.org/wsn/bw-2" xmlns:wsrf-rw="http://docs.oasis-open.org/wsrf/rw-2" xmlns:wsaw="http://www.w3.org/2006/05/addressing/wsdl" xmlns:wsrf-r="http://docs.oasis-open.org/wsrf/r-2" xmlns:trc="http://www.onvif.org/ver10/recording/wsdl" xmlns:tse="http://www.onvif.org/ver10/search/wsdl" xmlns:trp="http://www.onvif.org/ver10/replay/wsdl" xmlns:tnsn="http://www.eventextension.com/2011/event/topics"><env:Body><env:Fault><env:Code><env:Value>env:Sender</env:Value>
<env:Subcode><env:Value>ter:NotAuthorized</env:Value>
</env:Subcode>
</env:Code>
<env:Reason><env:Text xml:lang="en">The action requested requires authorization and the sender is not authorized</env:Text>
</env:Reason>
</env:Fault>
</env:Body>
</env:Envelope>

**Comment by bgilmer77** [comment]
I decided to run another test with things set exactly as you asked.  Here is the result of a log:list command
```
openhab> log:list
Logger                                             │ Level
───────────────────────────────────────────────────┼──────
ROOT                                               │ INFO
javax.jmdns                                        │ INFO
javax.mail                                         │ INFO
org.apache.karaf.internal.Kar.Service.Impl         │ INFO
org.apache.karaf.internal.KarServiceImpl           │ INFO
org.apache.karaf.jaas.modules.audit                │ INFO
org.apache.karaf.kar.internal.KarSerivceImpl       │ INFO
org.apache.karaf.kar.internal.KarServiceImpl       │ INFO
org.apache.karaf.kar.shell.ssh.SshUtils            │ INFO
org.apache.karaf.shell.ssh.SshUtils                │ INFO
org.apache.karaf.shell.support                     │ INFO
org.eclipse.lsp4j                                  │ OFF
org.eclipse.smarthome                              │ INFO
org.jupnp                                          │ INFO
org.openhab                                        │ INFO
org.openhab.binding.ipcamera                       │ DEBUG
org.openhab.ui.paper                               │ INFO
org.openhab.ui.paper.internal                      │ INFO
org.ops4j.pax.url.mvn.internal.AetherBasedResolver │ INFO
org.ops4j.pax.web.pax-web-runtime                  │ OFF
smarthome.event                                    │ INFO
smarthome.event.InboxUpdatedEvent                  │ INFO
smarthome.event.ItemAddedEvent                     │ INFO
smarthome.event.ItemRemovedEvent                   │ INFO
smarthome.event.ItemStateEvent                     │ INFO
smarthome.event.ThingAddedEvent                    │ INFO
smarthome.event.ThingRemovedEvent                  │ INFO
smarthome.event.ThingStatusInfoEvent               │ INFO
```
Here is the log file.

[openhab.ipcamDEBUG.othersINFO.log](https://github.com/Skinah/IpCamera/files/4685701/openhab.ipcamDEBUG.othersINFO.log)

**Comment by bgilmer77** [comment]
I do not think it makes a difference, but this is a dual-homed host.  I have one interface on the 192.168.18.0 network - this is my isolated camera network.  The other interface is on my control network, which is on 10.18.78.0.  I see a couple of DEBUG level error messages in the log, `Possible NIC/IP match found:192.168.18.5`.

**Comment by Skinah** [comment]
Please re-read my last reply. Your newer log still shows the same issue and until that is solved it is a waste of my time looking at your logs. The wireshark logs are impossible to come from the binding and must be because you have left onvif to mqtt scripts running.

**Comment by bgilmer77** [comment]
>Please re-read my last reply. Your newer log still shows the same issue and until that is solved it is a waste of my time looking at your logs.

Thank you for your time on this issue.  Absolutely will review your last reply.  Something strange is going on.  onvif2mqtt is not running as far as I know.  It is started from the command line (`npm run dev`), and killed with `CTRL-C`.  Maybe event subscription is persisted in the doorbell?  Upon PIR event, wireshark shows that doorbell shoots a short note to Intel NUC to start the exchange.  But there is still code on the OH2 server that is responding to the initial ping from the doorbell after a PIR event.  I will check more on this.

doorbell time should be pulling from NTP server running on OH2 Intel NUC.  But perhaps something is wrong with that.  

Won't contact you again until I have done more research.

**Comment by Skinah** [comment]
Usually the onvif messages keep getting sent until the camera is powered off and back on, or if the correct onvif message is sent asking to unsubscribe.

**Comment by bgilmer77** [comment]
Thanks @Skinah.  I did a little checking yesterday and what I am finding is very confusing.  It will take me a little bit of experimentation to figure out what is going on.
- Regarding time synchronization
I am aware that the times of the OH server and the cameras need to be the same for authorization to work.  Right from the beginning, I decided to set up my OH2 box as an NTP server for all my cameras.  The doorbell is set to sync with this server.  The OH2 server is set for Eastern Time, and the `timedatectl` NTP command verifies that this is the case.  When I use the HikVision Batch Configuration tool, it shows that the camera is set for Eastern time.  When I use the HikVision iOS app, it also shows that the camera is set for Eastern time.  When I use ODM, it shows the following `CST5:00:0004:00:00,M.3.2.0/02:00,M11.1.0/02:00:00`  I have no idea how to interpret this.  When I try to set the timezone using ODM, I get the error message, `Invalid data`.  ODM shows that the NTP server address is manually set to my OH2 server.  The on-screen display shows the same time as my OH2 server.
Given all of this, just to be sure, I am going to set the OH2 server to UTC, and try my best to set the camera to UTC as well and see what happens.
- Regarding the wireshark capture the other day where there appeared to be a dialog between the OH server and the doorbell during PIR events, I rebooted both the doorbell and the OH2 server.  Now, checking with wireshark, there is no dialog between the two during PIR events.  I started up the onvif2mqtt utility and then the two started exchanging information again.  I shut down the onvif2mqtt utility by entering `Ctrl-C`, and the exchange stopped, as you would expect.  Who knows - maybe I had some window open either locally or over an ssh session from another computer in my lab that was still running the onvif2mqtt utility.  To sum up, after rebooting, there is no dialog between the doorbell and OH2 after the initial "sender is not authorized" message during OH2 initialization.

So that is where I am for now.  I will do some careful testing and get back to you when I know more.

**Comment by bgilmer77** [comment]
**Update:**  Difference in clock times between OH server and camera does not seem to be causing the problem.  OH2 server and camera are both set to the same time.  The initial `NotAuthorized` message is caused by the binding issuing a `getProfiles` request without passing a username and nonce to the camera.  The sequence documented below occurs _without_ firing a PIR event.  This error occurs as the binding initiates communications with the camera at OH start.

I have additional information which I will put in a separate post.

**IpCamera version:** v5-30-2020

**Action:** Start OH2 server
**Result:** After ONVF `GetProfiles` request, response in OH log is `NotAuthorized`
**Cause:** `getProfiles` request is sent to camera without security.  I have verified that the command `getProfiles` functions properly if the appropriate nonce is passed in the XML as part of the request.

**Condition:**  Both OH server and doorbell camera set for UTC
**Verification of condition:**
- Real Time Clock of OH server verified by running `timedatectl`
```
root@OH-Test ~ # timedatectl
                      Local time: Mon 2020-06-01 21:08:53 UTC
                  Universal time: Mon 2020-06-01 21:08:53 UTC
                        RTC time: Mon 2020-06-01 21:08:53
                       Time zone: **UTC (UTC, +0000)**
       System clock synchronized: yes
systemd-timesyncd.service active: yes
                 RTC in local TZ: no
```

- Doorbell clock setting verified by running the curl command below:
```
curl http://192.168.18.30/onvif/device_service --data '<s:Envelope xmlns:s="http://www.w3.org/2003/05/soap-envelope"><s:Body xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xmlns:xsd="http://www.w3.org/2001/XMLSchema"><GetSystemDateAndTime xmlns="http://www.onvif.org/ver10/device/wsdl"/></s:Body></s:Envelope>' --header 'Content-Type: application/soap+xml; charset=utf-8; action="http://www.onvif.org/ver10/device/wsdl/GetSystemDateAndTime";'
```

*Doorbell returned 2020-06-01 21:08:53 UTC in XML as a response to the GetSystemDateAndTime command above*
```
<?xml version="1.0" encoding="UTF-8"?>
<env:Envelope xmlns:env="http://www.w3.org/2003/05/soap-envelope" xmlns:soapenc="http://www.w3.org/2003/05/soap-encoding" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xmlns:xs="http://www.w3.org/2001/XMLSchema" xmlns:tt="http://www.onvif.org/ver10/schema" xmlns:tds="http://www.onvif.org/ver10/device/wsdl" xmlns:trt="http://www.onvif.org/ver10/media/wsdl" xmlns:timg="http://www.onvif.org/ver20/imaging/wsdl" xmlns:tev="http://www.onvif.org/ver10/events/wsdl" xmlns:tptz="http://www.onvif.org/ver20/ptz/wsdl" xmlns:tan="http://www.onvif.org/ver20/analytics/wsdl" xmlns:tst="http://www.onvif.org/ver10/storage/wsdl" xmlns:ter="http://www.onvif.org/ver10/error" xmlns:dn="http://www.onvif.org/ver10/network/wsdl" xmlns:tns1="http://www.onvif.org/ver10/topics" xmlns:tmd="http://www.onvif.org/ver10/deviceIO/wsdl" xmlns:wsdl="http://schemas.xmlsoap.org/wsdl" xmlns:wsoap12="http://schemas.xmlsoap.org/wsdl/soap12" xmlns:http="http://schemas.xmlsoap.org/wsdl/http" xmlns:d="http://schemas.xmlsoap.org/ws/2005/04/discovery" xmlns:wsadis="http://schemas.xmlsoap.org/ws/2004/08/addressing" xmlns:xop="http://www.w3.org/2004/08/xop/include" xmlns:wsnt="http://docs.oasis-open.org/wsn/b-2" xmlns:wsa="http://www.w3.org/2005/08/addressing" xmlns:wstop="http://docs.oasis-open.org/wsn/t-1" xmlns:wsrf-bf="http://docs.oasis-open.org/wsrf/bf-2" xmlns:wsntw="http://docs.oasis-open.org/wsn/bw-2" xmlns:wsrf-rw="http://docs.oasis-open.org/wsrf/rw-2" xmlns:wsaw="http://www.w3.org/2006/05/addressing/wsdl" xmlns:wsrf-r="http://docs.oasis-open.org/wsrf/r-2" xmlns:trc="http://www.onvif.org/ver10/recording/wsdl" xmlns:tse="http://www.onvif.org/ver10/search/wsdl" xmlns:trp="http://www.onvif.org/ver10/replay/wsdl" xmlns:tnsn="http://www.eventextension.com/2011/event/topics"><env:Body><tds:GetSystemDateAndTimeResponse><tds:SystemDateAndTime><tt:DateTimeType>NTP</tt:DateTimeType>
<tt:DaylightSavings>false</tt:DaylightSavings>
<tt:TimeZone><tt:TZ>CST0:00:00</tt:TZ>
</tt:TimeZone>
<tt:UTCDateTime><tt:Time><tt:Hour>21</tt:Hour>
<tt:Minute>24</tt:Minute>
<tt:Second>5</tt:Second>
</tt:Time>
<tt:Date><tt:Year>2020</tt:Year>
<tt:Month>6</tt:Month>
<tt:Day>1</tt:Day>
</tt:Date>
</tt:UTCDateTime>
<tt:LocalDateTime><tt:Time><tt:Hour>21</tt:Hour>
<tt:Minute>24</tt:Minute>
<tt:Second>5</tt:Second>
</tt:Time>
<tt:Date><tt:Year>2020</tt:Year>
<tt:Month>6</tt:Month>
<tt:Day>1</tt:Day>
</tt:Date>
</tt:LocalDateTime>
</tds:SystemDateAndTime>
</tds:GetSystemDateAndTimeResponse>
</env:Body>
</env:Envelope>
```

**Comment by Skinah** [comment]
Thanks for looking into it and finding the root cause. I have begun replacing the ONVIF lib that I use with my own functions as there are two issues plus your third finding of how the library is breaking things. By replacing the functions I will reduce the dependancies down to just the Netty lib and we should see speed increases. This may take me months to finally have all features replaced but should be less to solve what your asking.

**Comment by bgilmer77** [comment]
@Skinah that is great news.  It turns out that this camera _does_ support onvif queries for streaming URLs as well.  I will post something on that in a few days.

If you want something to look at, it turns out that [the agsh/onvif library](https://github.com/agsh/onvif) contains an example (example6.js) that connects to a camera and executes a `CreatePullPointSubscription` and then reports events.  The example is handy in that it also 
executes a `GetEventProperties` command which provides a long list of event properties (turns out there are a lot!).

I would like to make a donation to your effort.  If I sent you one of these doorbells, would you have any use for it?  I would just buy it and have it shipped to you.  Let me know if you would like me to do that.

**Comment by bgilmer77** [comment]
Don't know if this is helpful...
```
ONVIF Support - getCapabilities

Reported directly from EZViz DB-1 running HikVision FW with no SD card installed
Command issued to camera - getCapabilities
Results captured as XML using wireshark

XAddr -
http://cam.ip.address/analytics

    RulesSupport - true
    AnalyticsModuleSupport - true

Xaddr -
http://cam.ip.address/onvif/device_service
Network

    IPFilter - true
    ZeroConfiguration - true
    IPVersion6 - false
    DynDNS - true
    Dot11Configuration - false
    Dot1x Configurations - 0

System

    DiscoveryResolve - false
    DiscoveryBye - true
    RemoteDiscovery - true
    SystemBackup - false
    SystemLogging - false
    FirmwareUpgrade - true
    SupportedVersions Major - 2
    SupportedVersions Minor - 60
    SupportedVersions Major - 2
    SupportedVersions Minor - 40
    SupportedVersions Major - 2
    SupportedVersions Minor - 20
    SupportedVersions Major - 2
    SupportedVersions Minor - 10
    SupportedVersions Major - 2
    SupportedVersions Minor - 0
    HttpFirmwareUpgrade - true
    HttpSystemBackup - false
    HttpSystemLogging - false
    HttpSupportInformation - false

IO

    InputConnectors - 0
    RelayOutputs - 0
    Auxiliary - false
    AuxiliaryCommands - nothing

Security

    TLS1.1 - false
    TLS1.2 - false
    OnboardKeyGeneration - false
    AccessPolicyConfig - false
    X.509Token - false
    SAMLToken - false
    KerbosToken - false
    RELToken - false
    TLS1.0 - false
    Dot1X - false
    SupportedEAPMethod - 0
    RemoteUserHandling - false

Events
Xaddr
http://cam.ip.address/onvif/Events

    WSSubscriptionPolicySupport - true
    WSPullPointSupport - true
    WSPausableSubscriptionManagerInterfaceSupport - false

Imaging
XAddr:
http://cam.ip.addr/onvif/Imaging

Media
XAddr:
http://cam.ip.addr/onvif/Media
StreamingCapabilities

    RTPMulticast - true
    RTP_TCP - true
    RTP_RTSP_TCP - true

ProfileCapabilities
- MaximumNumberOfProfiles - 10
DeviceIO
XAddr:
http://cam.ip.addr/onvif/DeviceIO

    VideoSources - 1
    VideoOutputs - 0
    AudioSources - 1
    AudioOutputs - 1
    RelayOutputs - 0

Recording
XAddr:
http://cam.ip.addr/onvif/Recording

    ReceiverSource - false
    MediaProfileSource - true
    DynamicRecordings - false
    DynamicTracks - false
    MaxStringLength - 64

Search
Xaddr:
http://cam.ip.addr/onvif/SearchRecording
- MetadataSearch - false
Replay
Xaddr:
http://cam.ip.addr/onvif/Replay
```

**Comment by Skinah** [comment]
Yes I could use one for testing in a white colour but I wish to stress that is not necessary but would be useful.

I have my own Netty based Onvif successfully connecting and returning password protected replies from all of my cameras so the main hurdle is now "jumped". Should be a matter of days to weeks to have a build to share. I don't need any traces unless the next build using a whole new set of methods does not work.

**Comment by Skinah** [comment]
Newer build has been made that may work for you, a lot of changes were made.

**Comment by bgilmer77** [comment]
Thanks @Skinah.  I am out of town but will see if I can try it remotely.  If not, I will get back to you next weekend.

**Comment by bgilmer77** [comment]
[openhab.log](https://github.com/Skinah/IpCamera/files/4753581/openhab.log)
So I had a chance to give it a run.  Certainly it is pulling a lot more info using ONVIF.

Brad

**Comment by Skinah** [comment]
Looks like it got a shutdown command only
Part way through connecting to the camera. Log did not show much. Any idea why it shutdown part way through?

**Comment by bgilmer77** [comment]
That was me.  I started OH, added the Thing file, let it run until the HTTP dialogs stopped, and then shutdown OH.  If you would like me to do something else, let me know please.

**Comment by bgilmer77** [comment]
The wireshark trace shows a 400 Bad Request - authentication required message, along with a 400 Bad Request returned for the GetProfiles request.

**Comment by bgilmer77** [comment]
I re-ran the test, starting OH with the thing file active.  Here is a complete wireshark capture of the communications between OH and the doorbell.  You can see the highlighted GetProfiles request with a 400 Bad Request response immediately below.

<img width="1283" alt="image" src="https://user-images.githubusercontent.com/7648629/84201063-b43ccb00-aa75-11ea-88b5-cf90cd70d242.png">

  Here are the last three lines of the log.
```
2020-06-09 21:10:55.031 [DEBUG] [nding.ipcamera.onvif.OnvifConnection] - We hit an issue parsing url:
2020-06-09 21:10:55.032 [DEBUG] [nding.ipcamera.onvif.OnvifConnection] - Sending ONVIF request:GetProfiles
2020-06-09 21:10:55.679 [INFO ] [org.quartz.core.QuartzScheduler     ] - Scheduler openHAB-job-scheduler_$_NON_CLUSTERED started.
```
OH is still running.  I just asked for a snapshot from the camera, and it produced one and added a new line in the log as follows.
```
2020-06-09 21:25:53.978 [DEBUG] [pcamera.internal.StreamServerHandler] - Stream Server recieved request 	GET:/ipcamera.jpg
```

So obviously OH did not hang.

**Comment by Skinah** [comment]
I only need the openhab log. Here’s an example of what it looks like.

https://community.openhab.org/t/ipcamera-new-ip-camera-binding/42771/1708

**Comment by Skinah** [comment]
You should not need to pull out entries as doing that you may remove something I need. Probably the best idea is to set everything to INFO level which is the default and I never touch them from the default settings.

Then set the binding to TRACE level.
Then restart the binding only using the karaf console.

This way the logs won’t fill up with other bindings info which may not hide passwords or other data.

I suspect there is a problem early on in the onvif connecting that needs sorting out first.

The bindings logs make things easier for me as they give feedback on the data they scrape and if something went wrong, this is why the logs are far more important then wireshark traces.

**Comment by bgilmer77** [comment]
Okay - log detail set as you requested.
-  started OH
-  stopped binding only using karaf console
-  started binding only using karaf console
-  stopped OH

See attached

[openhab.log](https://github.com/Skinah/IpCamera/files/4765859/openhab.log)

**Comment by Skinah** [comment]
Checked the dates in your log and they are from 17 days ago plus it is clear it is the wrong /old binding being used so that means it is a log from over two weeks ago.

Nothing can be used from that log and every openhab service was in debug mode, not just the one single binding.

**Comment by Skinah** [comment]
Are u using openhabian? Do you have frontail installed?

**Comment by bgilmer77** [comment]
Correct - old log.  I am still out of town connecting using a slow virtual desktop session.  Clearly I picked up the wrong file.  Try this one.

[openhab.log](https://github.com/Skinah/IpCamera/files/4770233/openhab.log)

Not running openhabian - running OH2 version 2.5.5 Release Build.  Do not have frontail installed.

**Comment by Skinah** [comment]
Log was good, thank you...
Have created a new build for you to try after comparing the wireshark logs to what is different in the binding I did make a change to make it closer...
Seems to be an authorising issue still, but the good news is that I now have all 3 of my cameras working with the events here, so it should be getting very close.

**Comment by bgilmer77** [comment]
Thanks for the new build.  I tried it.  Log below.
-  Deleted openhab.log and created a new log file
[openhab.log](https://github.com/Skinah/IpCamera/files/4771250/openhab.log)

-  Started OH
-  Stopped binding
-  Restarted binding
-  Stopped binding
-  Stopped Openhab

**Comment by Skinah** [comment]
New build just uploaded, only made a single header change and now they are identical to what ODM sends, if it does not work then it will need a wireshark trace done and compare the two to see what exactly is different on the 'GetProfiles' request that makes it fail.

Hikvision ones with API
Amcrest
Instar
Reolink

They are now all working with onvif events so if this build fails then I need to take a break and come back to it with fresh eyes.

**Comment by bgilmer77** [comment]
Thanks @Skinah.  Perhaps it is me?  Now it is throwing a java exception, "Duplicate channels ipcamera:ONVIF:doorbell:audioAlarm".  Attached is the log.

[openhab.log](https://github.com/Skinah/IpCamera/files/4776929/openhab.log)

**Comment by Skinah** [comment]
I have three different brand cameras and they all work fine as onvif thing type. Also the code is the same, the only difference is the api cameras skip setting up events but they still send the same initial onvif that yours is having an issue with.

Please check for password typos. Also how are you adding the cameras with thing files? If so what contents? Or is it paper ui inbox and manually or auto searching?

**Comment by Skinah** [comment]
Made a new build ipcamera-2020-07-05-Beta

Changes to Onvif which may help cameras that wont connect if they break the HTTP convention and are case sensitive to headers. This may be worth you trying this build.

**Comment by Skinah** [comment]
closing this down as the camera is now working well with the latest build. If there are any bugs feel free to open a new issue.
